osquery/tables/system/linux/services.cpp

```diff
@@ -0,0 +1,774 @@
+#include <boost/process.hpp>
+
+#include <osquery/core.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+/*
+  You can compile this with the system libraries by uncommenting these lines
+  and removing "services.h"
+
+  #include <systemd/sd-bus.h>
+  #include <systemd/sd-daemon.h>
+  #include <systemd/sd-login.h>
+*/
+
+#include "services.h"
+using namespace systemd_deps;
+
+namespace osquery {
+namespace {
+
+namespace boostfs = boost::filesystem;
+namespace boostproc = boost::process;
+
+bool sysVinitEnabled() {
+  static boost::optional<bool> enabled;
+  if (enabled != boost::none) {
+    return enabled.get();
+  }
+
+  static std::mutex m;
+  std::lock_guard<std::mutex> lock(m);
+  if (enabled != boost::none) {
+    return enabled.get();
+  }
+
+  enabled = boostfs::exists("/etc/init.d");
+  VLOG(1) << "SysVinit was" << (enabled.get() ? " " : " NOT ") << "found";
+
+  return enabled.get();
+}
+
+bool upstartEnabled() {
+  static boost::optional<bool> enabled;
+  if (enabled != boost::none) {
+    return enabled.get();
+  }
+
+  static std::mutex m;
+  std::lock_guard<std::mutex> lock(m);
+  if (enabled != boost::none) {
+    return enabled.get();
+  }
+
+  enabled = boostfs::exists("/etc/init");
+  VLOG(1) << "Upstart was" << (enabled.get() ? " " : " NOT ") << "found";
+
+  return enabled.get();
+}
+
+Status enumerateSysVinitServices(QueryData& query_data) {
+  if (!sysVinitEnabled()) {
+    return Status(0, "OK");
+  }
+
+  static const std::vector<std::string> run_level_dir_list = {"/etc/rc0.d",
+                                                              "/etc/rc1.d",
+                                                              "/etc/rc2.d",
+                                                              "/etc/rc3.d",
+                                                              "/etc/rc4.d",
+                                                              "/etc/rc5.d",
+                                                              "/etc/rc6.d",
+                                                              "/etc/rcS.d"};
+
+  std::unordered_map<std::string, std::set<size_t>> service_list;
+
+  for (size_t run_level = 0U; run_level < run_level_dir_list.size();
+       ++run_level) {
+    const auto& run_level_path = run_level_dir_list[run_level];
+
+    try {
+      boostfs::directory_iterator end_it;
+      for (auto it = boostfs::directory_iterator(run_level_path); it != end_it;
+           ++it) {
+        const auto& symlink = it->path();
+        if (!boostfs::is_symlink(symlink)) {
+          continue;
+        }
+
+        auto symlink_destination = boostfs::canonical(symlink);
+        if (!boostfs::is_regular_file(symlink_destination)) {
+          VLOG(1) << "Skipping broken symlink: " << symlink << " -> "
+                  << symlink_destination;
+
+          continue;
+        }
+
+        auto service_name = symlink_destination.filename().string();
+        service_list[service_name].insert(run_level);
```

Comment:
Set is more intuitive, but not necessary in that case, is not it?

Reply:
Yes, it's more intuitive, but the waste of CPU. You can not really duplicate anything, vector would do the job, will not it?